src/service.js

```diff
@@ -443,6 +498,35 @@ function getServiceInternal(holder, context, id, opt_factory,
   return s.obj;
 }
 
+/**
+ * @param {!Object} holder Object holding the service instance.
+ * @param {!Window|!./service/ampdoc-impl.AmpDoc} context Win or AmpDoc.
+ * @param {string} id of the service.
+ * @param {?function(new:Object, ?)=} opt_ctor
+ *     Constructor function to new the service. Called with context.
+ * @param {?function(?)=} opt_factory
+ *     Factory function to create the new service. Called with context.
+ */
+function registerServiceInternal(holder, context, id, opt_factory, opt_ctor) {
+  dev().assert(!!opt_factory != !!opt_ctor,
+    `Provide a constructor or a factory, but not both for service ${id}`);
+  const services = getServices(holder);
+  let s = services[id];
+  if (s && !s.obj && !s.ctor && !s.factory) {
```

Comment:
There is a slight difference here. The proposed code would allow a constructor or factory to be registered even if the other type of creator already exists.

For instance:

```
registerService(window, 'Foo', /* constructor */ undefined, window  => new  FooClass1(window));
registerService(window, 'Foo', FooClass2);
getService(window, 'Foo')
```
The last line will return an instance of FooClass2, when it should return an instance of FooClass1.